badger/badger.go

```diff
@@ -0,0 +1,346 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package badger
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"github.com/dgraph-io/badger/v3"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/go-stoabs/util"
+	"github.com/sirupsen/logrus"
+)
+
+var _ stoabs.ReadTx = (*tx)(nil)
+var _ stoabs.WriteTx = (*tx)(nil)
+var _ stoabs.Reader = (*badgerShelf)(nil)
+var _ stoabs.Writer = (*badgerShelf)(nil)
+
+const defaultFileTimeout = 5 * time.Second
+
+var fileTimeout = defaultFileTimeout
+
+// CreateBadgerStore creates a new Badger-backed KV Store.
+func CreateBadgerStore(filePath string, opts ...stoabs.Option) (stoabs.KVStore, error) {
+	cfg := stoabs.DefaultConfig()
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	badgerOpts := badger.DefaultOptions(filePath)
+	if cfg.NoSync {
+		badgerOpts = badgerOpts.WithInMemory(true).WithDir("").WithValueDir("")
+	}
+	if cfg.Log != nil {
+		// badgerOpts = badgerOpts.WithSyncWrites(true) // hard writes for single node usage
+		badgerOpts = badgerOpts.WithLogger(cfg.Log)
+	}
+
+	return createBadgerStore(filePath, badgerOpts, cfg)
+}
+
+func createBadgerStore(filePath string, options badger.Options, cfg stoabs.Config) (stoabs.KVStore, error) {
+	err := os.MkdirAll(path.Dir(filePath), 0644)
+	if err != nil {
+		return nil, err
+	}
+
+	db, err := badger.Open(options)
+	if err != nil {
+		return nil, err
+	}
+
+	return Wrap(db, cfg), nil
+}
+
+// Wrap creates a KVStore using an existing badger.DB
+func Wrap(db *badger.DB, cfg stoabs.Config) stoabs.KVStore {
+	return &Store{
+		DB:  db,
+		log: cfg.Log,
+	}
+}
+
+type Store struct {
+	DB    *badger.DB
+	log   *logrus.Logger
+	mutex sync.Mutex
+}
+
+func (b *Store) Close(ctx context.Context) error {
+	return util.CallWithTimeout(ctx, b.DB.Close, func() {
+		b.log.Error("Closing of Badger Store timed out, Store may not shut down correctly.")
+	})
+}
+
+func (b *Store) Write(ctx context.Context, fn func(stoabs.WriteTx) error, opts ...stoabs.TxOption) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		return fn(tx)
+	}, true, opts)
+}
+
+func (b *Store) Read(ctx context.Context, fn func(stoabs.ReadTx) error) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		return fn(tx)
+	}, false, nil)
+}
+
+func (b *Store) WriteShelf(ctx context.Context, shelfName string, fn func(writer stoabs.Writer) error) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		shelf, err := tx.GetShelfWriter(shelfName)
+		if err != nil {
+			return err
+		}
+		return fn(shelf)
+	}, true, nil)
+}
+
+func (b *Store) ReadShelf(ctx context.Context, shelfName string, fn func(reader stoabs.Reader) error) error {
+	return b.doTX(ctx, func(tx *tx) error {
+		shelf := tx.GetShelfReader(shelfName)
+		return fn(shelf)
+	}, false, nil)
+}
+
+func (b *Store) doTX(ctx context.Context, fn func(tx *tx) error, writable bool, opts []stoabs.TxOption) error {
+	if writable {
+		b.mutex.Lock()
```

Comment:
BBolt uses `util/ContextRWLocker` to support time-outs (acquisition), why not use it here as well?

Side-note: appearantely not tested for all databases (otherwise it would faiL)

Reply:
removed mutex, badger supports parallel TXs